feat: add childLoggerFactory config

brettwillis opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md

By making a contribution to this project, I certify that:

* (a) The contribution was created in whole or in part by me and I
  have the right to submit it under the open source license
  indicated in the file; or

* (b) The contribution is based upon previous work that, to the best
  of my knowledge, is covered under an appropriate open source
  license and I have the right under that license to submit that
  work with modifications, whether created in whole or in part
  by me, under the same open source license (unless I am
  permitted to submit under a different license), as indicated
  in the file; or

* (c) The contribution was provided directly to me by some other
  person who certified (a), (b) or (c) and I have not modified
  it.

* (d) I understand and agree that this project and the contribution
  are public and that a record of the contribution (including all
  personal information I submit with it, including my sign-off) is
  maintained indefinitely and may be redistributed consistent with
  this project or the open source license(s) involved.
-->

Allow customisation of child logger bindings (or the instance itself) via child logger factory function.

Closes #4751 

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

brettwillis commented:
Referring to https://github.com/fastify/fastify/issues/4751#issuecomment-1548582144, not sure if we should also do a `fastify.setChildLoggerFactory` instance method to allow it to be set/encapsulated by a plugin?

If so, not sure if there's anything special I have to do to allow encapsulation etc...

metcoder95 commented:
On my side, I think it can be an interesting addition, especially for plugins to have capabilities to also maybe decorate the logger with extra properties (e.g. `traceId`, custom headers, etc.)

Though, I would like to hear from @fastify/core their thoughts 🙂

metcoder95 commented on lib/logger.js:
```diff
@@ -107,27 +107,64 @@ function createLogger (options) {
+  const loggerBindings = {
+    [options.requestIdLogLabel]: reqId
+  }
+  const child = options.childLoggerFactory(logger, loggerBindings, loggerOpts || {}, req)
```
Wondering if the `childLoggerFactory` should have access to the `this` context linked to the `fastify` instance 🤔

brettwillis replied:
I was thinking so too, however I just haven't understood the code enough to know how to reference the fastify instance in these 3 files. 😓 

Furthermore with the typings, to strictly/properly specify the 5 generic parameters of the `FastifyInstance` type, I'd need to add `RawRequest`, `RawReply`, `TypeProvider` generic params to the `FastifyServerOptions` type which may be a breaking change? Otherwise we leave those generic parameters at their default value which would also work, although not type-correct in the strictest sense.

brettwillis replied:
> Wondering if the childLoggerFactory should have access to the this context linked to the fastify instance

Ok I bound `this` to the Fastify instance. In some cases (404) is is the root instance. In normal cases it is the encapsulated instance (I think).

Really need an export to check this last commit very carefully because I'm not 100% confident that I've referenced the correct Fastify instance in each case.

brettwillis replied:
@mcollina did you carefully check that I am referencing the correct fastify instance in each case?

metcoder95 replied:
>Really need an export to check this last commit very carefully because I'm not 100% confident that I've referenced the correct Fastify instance in each case.

See my previous comment, you did it right, but better to fallback into the instance owner of the `forOhFour` 🙂

mcollina commented:
We use tsd for testing the types

brettwillis commented:
Added the type check in the last commit

mcollina reviewed (approved):
lgtm

metcoder95 commented on lib/fourOhFour.js:
```diff
@@ -64,7 +66,7 @@ function fourOhFour (options) {
     return function onBadUrl (path, req, res) {
       const id = genReqId(req)
-      const childLogger = logger.child({ reqId: id })
+      const childLogger = createChildLogger(this, logger, req, id, options)
```
For this scenario, given the nature of the `forOhFour`, let's fallback inoto the `kFourOhFourLevelInstance` which directly points to the instance owner of the 404 handler

metcoder95 replied:
You're right, missed that 👍

metcoder95 commented on lib/fourOhFour.js:
```diff
@@ -172,7 +174,7 @@ function fourOhFour (options) {
     // we can
     const id = genReqId(req)
-    const childLogger = logger.child({ reqId: id })
+    const childLogger = createChildLogger(this, logger, req, id, options)
```
Same as my previous comment

metcoder95 commented:
>If we want to add fastify.setChildLoggerFactory() for encapsulation, then I need some guidance. Really not sure where to begin so need some hints. Do I just add the method to the root fastify object returned from the constructor?

To achieve this we need to do the following:
1. We will need to define a default `childLoggerFactory` fn when instantiating [`fastify`](https://github.com/fastify/fastify/blob/e82f4a7d463e49c5b9b33011a590979e872c4571/fastify.js#L226). We need hide the child logger fn behind a symbol.
2. We need to define it within the [`Context`](https://github.com/fastify/fastify/blob/e82f4a7d463e49c5b9b33011a590979e872c4571/lib/context.js#L21) object. It should use the one attached to `fastify` at the previous step using the symbol.
3. The `fastify.setChildLoggerFactory` should modify the default child logger fn hidden with the symbol, so the `context` object can reference it within nested encapsulations. e.g. [`setSerializerCompiler`](https://github.com/fastify/fastify/blob/main/fastify.js#L754)

brettwillis commented:
Ok I added encapsulation per-plugin and per-route, like `errorHandler` etc. And it looks like I've done it right because I added tests similar to `errorHandler` and it seems to be working. Let me know your thoughts.

My only note is: configuration options like `errorHandler` which have a `setErrorHandler()` funciton on the fastify instance don't currently also have a constructor option. However for the child logger factory we have both the `childLoggerFactory` constructor option, and the `setChildLoggerFactory` option. Do we remove the constructor option now or keep both?

metcoder95 commented on lib/route.js:
```diff
@@ -414,7 +413,7 @@ function buildRouting (options) {
       loggerOpts.serializers = context.logSerializers
     }
-    const childLogger = createChildLogger(context.server, logger, req, id, childLoggerOpts, loggerOpts)
+    const childLogger = createChildLogger(context.server, logger, req, id, context, loggerOpts)
```
Maybe we can pass the whole context instead of the server and later the whole context? 
```suggestion
    const childLogger = createChildLogger(context, logger, req, id, loggerOpts)
```

metcoder95 replied:
Do you mean behaviour over the `createChildLogger` function?

metcoder95 commented on test/encapsulated-child-logger-factory.test.js:
```diff
@@ -0,0 +1,39 @@
+const { test } = require('tap')
+const Fastify = require('..')
+
+test('encapsulates an child logger factory', async t => {
```
Can we add one for testing when registering a plugin (not creating another child context but rather altering the parent)? When using `fastify-plugin`

metcoder95 replied:
Yes, exactly what you suggested 👍

brettwillis replied:
Done

metcoder95 commented on types/instance.d.ts:
```diff
@@ -542,6 +542,34 @@ export interface FastifyInstance<
     handler: (this: FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider>, error: TError, request: FastifyRequest<RouteGeneric, RawServer, RawRequest, SchemaCompiler, TypeProvider>, reply: FastifyReply<RawServer, RawRequest, RawReply, RouteGeneric, ContextConfigDefault, SchemaCompiler, TypeProvider>) => any | Promise<any>
   ): FastifyInstance<RawServer, RawRequest, RawReply, Logger, TypeProvider>;
 
+  /**
```
Maybe we should to provide a standalone type as `FastifyChildLoggerFactory`, wdyt?

metcoder95 reviewed (commented):
>Do we remove the constructor option now or keep both?

You're right. I think that only having the one through the method is enough, but happy to hear other thoughts 🙂

brettwillis commented:
So sorry guys, I posted a whole bunch of comments several weeks ago and didn't realise I had to submit the review for them to appear 🤦🏼 

Here is what is pending review

1. @metcoder95 a need your clarification here (https://github.com/fastify/fastify/pull/4760#discussion_r1206962134)
2. @metcoder95 I implemented your suggestion (https://github.com/fastify/fastify/pull/4760#discussion_r1206958652) but it had a couple of implications (https://github.com/fastify/fastify/pull/4760#discussion_r1207570420, https://github.com/fastify/fastify/pull/4760#discussion_r1207572291) which I pointed out which seem benign but need you to confirm.
3. Need consensus on whether, now that we have added instance method `setChildLoggerFactory`, we should remove the `childLoggerFactory` constructor option or not.

metcoder95 commented on fastify.js:
```diff
@@ -713,7 +724,7 @@ function fastify (options) {
         if (frameworkErrors) {
           const id = genReqId(req)
-          const childLogger = logger.child({ reqId: id })
+          const childLogger = createChildLogger(onBadUrlContext, logger, req, id)
```
It was because the `badUrl` is handled differently than other cases. For instance, it creates the `Request` and `Reply` objects if the `frameworkErrors` hanlder is passed when creating the fastify app.

It shouldn't have negative impact, but we can ask @mcollina for a second pair of eyes to verify

metcoder95 commented on lib/fourOhFour.js:
```diff
@@ -172,7 +167,8 @@ function fourOhFour (options) {
     // we can
     const id = genReqId(req)
-    const childLogger = logger.child({ reqId: id })
+    const fourOhFourContext = this[kFourOhFourLevelInstance][kFourOhFourContext]
```
This has the same goal as https://github.com/fastify/fastify/pull/4760#discussion_r1236584946 but for the `onBadUrlHandler` instead

metcoder95 commented:
> So sorry guys, I posted a whole bunch of comments several weeks ago and didn't realise I had to submit the review for them to appear 🤦🏼
> 
> Here is what is pending review
> 
> 1. @metcoder95 a need your clarification here ([Add childLoggerFactory config #4760 (comment)](https://github.com/fastify/fastify/pull/4760#discussion_r1206962134))

Done ✅ 

> 2. @metcoder95 I implemented your suggestion ([Add childLoggerFactory config #4760 (comment)](https://github.com/fastify/fastify/pull/4760#discussion_r1206958652)) but it had a couple of implications ([Add childLoggerFactory config #4760 (comment)](https://github.com/fastify/fastify/pull/4760#discussion_r1207570420), [Add childLoggerFactory config #4760 (comment)](https://github.com/fastify/fastify/pull/4760#discussion_r1207572291)) which seem benign but need you to confirm.

✅ from my side

> 3. Need consensus on whether, now that we have added instance method `setChildLoggerFactory`, we should remove the `childLoggerFactory` constructor option or not.

Thinking twice, I feel that we can have both implemented. In that sense, we enable situations where we want to maintain a standard logger factory across the app, and just be overwritten by the context where we set a plugin to do so, affecting just a subgraph of the application, wdyt?

brettwillis commented:
> I'm +1 on removing the one from the constructor and stick to the method from the instance. In that manner works as we have for setNotFoundHandler and others 

Ok I removed the constructor option.

brettwillis commented on test/inject.test.js:
Ok so it seems like the changes I make to `onBadUrlContext` (making it a proper context) has made some of the recent #4826 obsolete.

That PR was to fix parts of the context being undefined because it wasn't a proper context, now it is a proper context.

brettwillis replied:
I don't think it's an issue. Just pointing it out. I needed to remove that test that expects `routeConfig` to be `undefined` as it is no longer undefined.

metcoder95 commented on test/inject.test.js:
```diff
@@ -458,14 +458,14 @@ test('should handle errors in builder-style injection correctly', async (t) => {
       t.ok(req.raw instanceof Readable)
       t.same(req.routerPath, undefined)
-      t.same(req.routeConfig, undefined)
+      // t.same(req.routeConfig, undefined)
```
Left over?

brettwillis replied:
Yeah it seems like the changes I made to onBadUrlContext (making it a proper context) has made some of the recent https://github.com/fastify/fastify/pull/4826 obsolete (it seems part of that PR was to fix parts of the context being undefined because it wasn't a proper context, now it is a proper context).

This test was added in that PR (to assert the context is not a full context?), but not it is a full context so that test fails and I had to remove it.

Shall I to remove it rather than comment it?

metcoder95 replied:
It was mostly to test that it does not throw when the `req.routeConfig` is `undefined`, although now is not the case. It should be fine just removing it 👍

brettwillis replied:
Done

metcoder95 reviewed (commented):
Overall, LGTM 👍 
Just about some left over and a conflict to be solved, and ready to go

jsumners commented on docs/Reference/Routes.md:
```diff
@@ -90,6 +90,14 @@ fastify.route(options)
   that this will point to fastify's default `errorHandler` only if a plugin
   hasn't overridden it already.
+* `childLoggerFactory(logger, binding, opts, rawReq)`: a custom factory function
+  that will be called to product a child logger instance for every request.
```
```suggestion
  that will be called to produce a child logger instance for every request.
```

metcoder95 commented:
Linter seems failing 🙁

brettwillis commented:
> Linter seems failing 🙁

Seems to be on files I didn't touch, but another PR #4879 . Do you want me to fix it here or you do somewhere else?

metcoder95 commented:
>Seems to be on files I didn't touch, but another PR https://github.com/fastify/fastify/pull/4879 . Do you want me to fix it here or you do somewhere else?

Ha, interesting. Didn't know how it passed the checks, let's open a new PR for it and rebase this branch. Opened #4882.

metcoder95 commented:
Are we good to merge it?

cc: @mcollina

brettwillis commented:
Thanks team and especially @metcoder95 for your support on this one 🙂
